client/web/src/cody/management/subscription/new/CodyProCheckoutForm.tsx

```diff
@@ -14,99 +18,47 @@ export const CodyProCheckoutForm: React.FunctionComponent<{
     stripePromise: Promise<Stripe | null>
     customerEmail: string | undefined
 }> = ({ stripePromise, customerEmail }) => {
-    const [clientSecret, setClientSecret] = useState('')
-    const [errorDetails, setErrorDetails] = useState('')
-    const [urlSearchParams] = useSearchParams()
-
-    // Optionally support the "showCouponCodeAtCheckout" URL query parameter, which is present
+    // Optionally support the "showCouponCodeAtCheckout" URL query parameter, which if present
     // will display a "promotional code" element in the Stripe Checkout UI.
+    const [urlSearchParams] = useSearchParams()
     const showPromoCodeField = urlSearchParams.get('showCouponCodeAtCheckout') !== null
 
-    // Issue an API call to the backend asking it to create a new checkout session.
-    // This will update clientSecret/errorDetails asynchronously when the request completes.
-    useEffect(() => {
-        // useEffect will not accept a Promise, so we call
-        // createCheckoutSession and let it run async.
-        // (And not `await createCheckoutSession` or `return createCheckoutSession`.)
-        void createCheckoutSession('monthly', showPromoCodeField, customerEmail, setClientSecret, setErrorDetails)
-    }, [customerEmail, showPromoCodeField, setClientSecret, setErrorDetails])
+    const req: CreateCheckoutSessionRequest = {
+        interval: 'monthly',
+        seats: 1,
+        customerEmail,
+        showPromoCodeField,
 
-    const options /* unexported EmbeddedCheckoutProviderProps.options */ = {
-        clientSecret,
+        // URL the user is redirected to when the checkout process is complete.
+        //
+        // BUG: Due to the race conditions between Stripe, the SSC backend,
+        // and Sourcegraph.com, immediately loading the Dashboard page isn't
+        // going to show the right data reliably. We will need to instead show
+        // some intersitular or welcome prompt, to give various things to sync.
+        returnUrl: `${origin}/cody/manage?session_id={CHECKOUT_SESSION_ID}`,
     }
+
+    // Make the API call to create the Stripe Checkout session.
+    const call = useMemo(() => Client.createStripeCheckoutSession(req), [])
```

Comment:
```suggestion
    const call = useMemo(() => Client.createStripeCheckoutSession(req), [req])
```

Same as https://github.com/sourcegraph/sourcegraph/pull/62715#discussion_r1603095355: we must include `req` as a hook dependency.
`req` object is re-cretated on each render. So either we memoize it too, e.g.
```
const req = useMemo(() => ({
        // ... other fields
        customerEmail,
        showPromoCodeField,
    }), 
    [customerEmail, showPromoCodeField] // reactive dependencies - when they change we want `req` value to be recalculated
)
```
or create `req` inside a memoized `call`, e.g.,
```
const call = useMemo(() =>
    Client.createStripeCheckoutSession({
        // ... other fields
        customerEmail,
        showPromoCodeField,
    }),
    [customerEmail, showPromoCodeField]
) 
```

I prefer the latter as `req` is only used once and there's no additional value in definitin it as a separate `const`.

Reply:
Good catch! Fixed. And yeah, I agree the second format where we construct the request object too as part of the memoization process is probably the easier to understand.